python/torch_mlir/csrc/base_lazy_backend/mlir_lowering_context.cpp

```diff
@@ -140,6 +141,32 @@ ComputationPtr TorchMlirLoweringContext::Build() {
       /*getArgAttribute=*/[](int) -> MlirAttribute { return {nullptr}; },
       /*importOptions=*/{/*assumeTensorsHaveValueSemantics=*/true});
 
+  MlirLocation loc = mlirLocationUnknownGet(mlir_context_);
+  MlirModule moduleOp = mlirModuleCreateEmpty(loc);
+  MlirBlock moduleBody = mlirModuleGetBody(moduleOp);
+  mlirBlockInsertOwnedOperation(moduleBody, 0, func_op);
+
+  TORCH_CHECK(
+      !mlirModuleIsNull(moduleOp),
+      "MlirModule is null.");
+
+  // Apply passes.
+  MlirPassManager pm = mlirPassManagerCreate(mlir_context_);
+  MlirOpPassManager nestedFuncPm = mlirPassManagerGetNestedUnder(
+    pm, mlirStringRefCreateFromCString("func.func"));
+
+  MlirPass refineTypesPass = mlirCreateRefineTypes();
```

Comment:
This breaks the architecture we are trying to maintain: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#overview

In particular, we should already know precise types in LTC and not need to run RefineTypes. Is there a way to fix this on the PyTorch side?

Reply:
I found that after we apply `Scalar::Lower` and create an operation with `Tensor` output type [here](https://github.com/llvm/torch-mlir/blob/main/python/torch_mlir/csrc/base_lazy_backend/mlir_node_lowering.cpp#L284), so then PyTorch [is trying to align types](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L131-L137) and insert `Int/Float/Complex/Scalar-ImpliIcit` ops. And in my case I get this graph:
```
  ...
  %1 : Long(requires_grad=0, device=cpu) = prim::Constant[value={1}]()
  %2 : Long(requires_grad=0, device=cpu) = prim::Constant[value={1}]()
  ...
  %23 : Scalar = aten::ScalarImplicit(%2)
  %24 : Scalar = aten::ScalarImplicit(%1)
  %25 : Half(2, 10) = aten::addmm(%22, %p2, %12, %23, %24)
```
Based on `aten::addmm` spec it expects `Scalar` type (NumberType) for the 4th and 5th inputs, so that is why we get `ScalarImplicit` instead of `Int/Float-Implicit`.

But I guess that during type alignment in PyTorch for the case when destination type is `NumberType` we can deduce precise type from `Tensor` like this:
```cpp
...
} else if (concrete_number) {
        auto tensorType = value->type()->cast<TensorType>();
        if (auto scalarType = tensorType->scalarType()) {
            switch (*scalarType) {
              // This is just an example. Need to add more types.
              case at::ScalarType::Byte:
              case at::ScalarType::Char:
              case at::ScalarType::Short:
              case at::ScalarType::Int:
              case at::ScalarType::Long:
                value = graph.insert(aten::IntImplicit, {value}, {}, loc);
                break;
              default:
                value = graph.insert(aten::ScalarImplicit, {value}, {}, loc);
            }
        } else {
          value = graph.insert(aten::ScalarImplicit, {value}, {}, loc);
        }
      }
...
```

Using this patch I was able to get correct JIT graph with refined types:
```
  ... 
  %1 : Long(requires_grad=0, device=cpu) = prim::Constant[value={1}]()
  %2 : Long(requires_grad=0, device=cpu) = prim::Constant[value={1}]()
  ... 
  %23 : int = aten::IntImplicit(%2)
  %24 : int = aten::IntImplicit(%1)
  %25 : Half(2, 10) = aten::addmm(%22, %p2, %12, %23, %24)
```
and torch-mlir:
```
  ...
  %0 = torch.vtensor.literal(dense<1> : tensor<si64>) : !torch.vtensor<[],si64>
  %1 = torch.vtensor.literal(dense<1> : tensor<si64>) : !torch.vtensor<[],si64>
 ...
  %7 = torch.aten.IntImplicit %1 : !torch.vtensor<[],si64> -> !torch.int
  %8 = torch.aten.IntImplicit %0 : !torch.vtensor<[],si64> -> !torch.int
  %9 = torch.aten.addmm %6, %arg2, %4, %7, %8 : !torch.vtensor<[10],f16>, !torch.vtensor<[2,16],f16>, !torch.vtensor<[16,10],f16>, !torch.int, !torch.int -> !torch.vtensor<[2,10],f16>
```

On the one hand there is nothing criminal in that we make type more precise, but on the other hand the `Scalar` type was also correct. So I think it is better do discuss this change with PyTorch team. 
Is there any chance that we keep `RefineTypes` transformation for now in base_lazy_backend until we push this fix to the PyTorch?

Reply:
Adding RefineTypes here significantly breaks the architecture of Torch-MLIR, so we do not want to add it, even temporarily.

Reply:
Okay, I'll try to make a fix on PyTorch side.
BTW, is there a list of requirements in PyTorch that the LTC satisfies? I mean guaranties for value semantics, type refinement etc.

Reply:
Okay, I'll try to make a fix on PyTorch side.
BTW, is there a list of requirements in PyTorch that the LTC satisfies? I mean guaranties for value semantics, type refinement etc.

Reply:
https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#the-backend-contract -- it should be checked by the SatisfiesBackendContract pass you added.

See also this for reference: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#lazytensorcore